File: src/mappers/__tests__/apple-music-json-mapper.test.ts
import { readFileSync } from 'fs';
import { join } from 'path';
import type { AppleMusicApiResource } from '../apple-music-json-mapper';
import { mapAlbum, mapArtist, mapSong } from '../apple-music-json-mapper';

function loadFixture(name: string): AppleMusicApiResource {
  const path = join(__dirname, '../../../fixtures', name);
  return JSON.parse(readFileSync(path, 'utf8')) as AppleMusicApiResource;
}

const catalogSong = loadFixture('catalog-song.json');
const catalogAlbum = loadFixture('catalog-album.json');
const libraryArtist = loadFixture('library-artist.json');

describe('apple-music-json-mapper', () => {
  it('maps catalog song fixtures', () => {
    expect(mapSong(catalogSong)).toEqual({
      id: '1441164424',
      title: 'Yesterday',
      artistName: 'The Beatles',
      artworkUrl: 'https://is1-ssl.mzstatic.com/image/thumb/Music/200x200bb.jpg',
      duration: '125000',
    });
  });

  it('maps catalog album fixtures', () => {
    expect(mapAlbum(catalogAlbum)).toEqual({
      id: '1441164425',
      title: 'Help!',
      artistName: 'The Beatles',
      artworkUrl: 'https://is1-ssl.mzstatic.com/image/thumb/Music/200x200bb.jpg',
      trackCount: '14',
    });
  });

  it('maps library artist fixtures', () => {
    expect(mapArtist(libraryArtist)).toEqual({
      id: 'l.abc123',
      name: 'The Beatles',
      artworkUrl: 'https://is1-ssl.mzstatic.com/image/thumb/Music/200x200bb.jpg',
    });
  });
});

File: android/src/test/java/expo/modules/applemusic/AppleMusicJsonMapperTest.kt
package expo.modules.applemusic

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Test

class AppleMusicJsonMapperTest {
  private fun loadFixture(name: String): JSONObject {
    val stream =
      checkNotNull(javaClass.classLoader?.getResourceAsStream("fixtures/$name")) {
        "Missing fixture: $name"
      }
    return JSONObject(stream.bufferedReader().readText())
  }

  @Test
  fun mapSong_catalogResource() {
    val resource = loadFixture("catalog-song.json")
    val song = AppleMusicJsonMapper.mapSong(resource)
    assertEquals("1441164424", song["id"])
    assertEquals("Yesterday", song["title"])
    assertEquals("The Beatles", song["artistName"])
    assertEquals("125000", song["duration"])
    assertEquals(
      "https://is1-ssl.mzstatic.com/image/thumb/Music/200x200bb.jpg",
      song["artworkUrl"],
    )
  }

  @Test
  fun mapAlbum_catalogResource() {
    val resource = loadFixture("catalog-album.json")
    val album = AppleMusicJsonMapper.mapAlbum(resource)
    assertEquals("1441164425", album["id"])
    assertEquals("Help!", album["title"])
    assertEquals("The Beatles", album["artistName"])
    assertEquals("14", album["trackCount"])
  }

  @Test
  fun mapArtist_libraryResource() {
    val resource = loadFixture("library-artist.json")
    val artist = AppleMusicJsonMapper.mapArtist(resource)
    assertEquals("l.abc123", artist["id"])
    assertEquals("The Beatles", artist["name"])
    assertEquals(
      "https://is1-ssl.mzstatic.com/image/thumb/Music/200x200bb.jpg",
      artist["artworkUrl"],
    )
  }
}
